### PR TITLE
Move CommandLineUtils to Common

### PR DESCRIPTION
**Pull request by BrennanConroy**
https://github.com/aspnet/dnx/issues/3246

### CODE REVIEW

File: src/Microsoft.Extensions.CommandLineUtils.Sources/CommandLine/AnsiConsole.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+
+namespace Microsoft.Dnx.Runtime.Common.CommandLine
```
Comment (pranavkm) [inline src/Microsoft.Extensions.CommandLineUtils.Sources/CommandLine/AnsiConsole.cs]: `Microsoft.Extensions....`

File: src/Microsoft.Extensions.CommandLineUtils.Sources/CommandLine/CommandLineApplication.cs [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+            RemainingArguments = new List<string>();
+            Invoke = () => 0;
+        }
+
+        public CommandLineApplication Parent { get; set; }
+        public string Name { get; set; }
+        public string FullName { get; set; }
+        public string Syntax { get; set; }
+        public string Description { get; set; }
+        public List<CommandOption> Options { get; private set; }
```
Comment (pranavkm) [inline src/Microsoft.Extensions.CommandLineUtils.Sources/CommandLine/CommandLineApplication.cs]: Drop the private setters for readonly properties.

File: src/Microsoft.Extensions.CommandLineUtils.Sources/CommandLine/CommandLineApplication.cs [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+        public CommandOption HelpOption(string template)
+        {
+            // Help option is special because we stop parsing once we see it
+            // So we store it separately for further use
+            OptionHelp = Option(template, "Show help information", CommandOptionType.NoValue);
+
+            return OptionHelp;
+        }
+
+        public CommandOption VersionOption(string template,
```
Comment (pranavkm) [inline src/Microsoft.Extensions.CommandLineUtils.Sources/CommandLine/CommandLineApplication.cs]: Indentation: just use VS defaults

**Comment by pranavkm** [comment]
If DNX isn't going to use this, could this be a non-sources package?

File: test/Microsoft.Extensions.CommandLineUtils.Sources.Tests/project.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        "xunit.runner.aspnet": "2.0.0-aspnet-*"
+    },
+
+    "frameworks": {
+        "dnx451": { },
+        "dnxcore50": { }
+    },
+
+    "commands": {
+        "run": "xunit.runner.aspnet",
```
Comment (pranavkm) [inline test/Microsoft.Extensions.CommandLineUtils.Sources.Tests/project.json]: Remove?

**Comment by pranavkm** [comment]
:watch: for reformatting \ namespace cleanup

**Comment by rynowak** [comment]
> If DNX isn't going to use this, could this be a non-sources package?

TestHost uses this btw.

**Comment by pranavkm** [comment]
@rynowak is it an issue for it to use a non-sources package? `sources` package are harder for external folks to consume (since they have to include the `aspnetmaster` feed).

**Comment by rynowak** [comment]
No, just making the point that it's going to be a breaking change that will need to fixed. Additionally, this starts introducing cycles if your product code has its own reference to this as a non-sources package. You could 'lift' the reference to something incompatible with TestHost and cause lots of fun issues.

If we're going to do that change we might just want to wait until we no longer own testhost (soon :+1:)

**Comment by BrennanConroy** [comment]
:up: :date:

File: src/Microsoft.Extensions.CommandLineUtils.Sources/CommandLine/CommandLineApplication.cs [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+                return VersionOption(template, () => shortFormVersion);
+            }
+            else
+            {
+                return VersionOption(template, () => shortFormVersion, () => longFormVersion);
+            }
+        }
+
+        // Helper method that adds a version option
+        public CommandOption VersionOption(string template,
```
Comment (pranavkm) [inline src/Microsoft.Extensions.CommandLineUtils.Sources/CommandLine/CommandLineApplication.cs]: More indentation.

**Comment by pranavkm** [comment]
The namespace is changing - so it's going to be a breaking change regardless. Might save us the trouble of doing it twice over if we do it now.

@BrennanConroy could you check with the powers to be to determine if we should make this a non-sources package? :shipit: once you have a resolution for it.

**Comment by Eilon** [comment]
Should people in general move to using the BCL command line parser? Will this one just die in the near future? Or do we significantly prefer it?

**Comment by pranavkm** [comment]
There's a BCL parser? We should switch rather than maintain additional code that's non-core.

**Comment by Eilon** [comment]
I thought there was. Someone mentioned it somewhere once upon a time, no? At that place. With the things...

**Comment by davidfowl** [comment]
Let's just make it a regular package right now. If we want to switch over to system.command line then well need to work on it as well.

File: src/Microsoft.Extensions.CommandLineUtils.Sources/project.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{
+  "version": "1.0.0-*",
+  "compilationOptions": {
+    "define": [ "TRACE" ],
+    "warningsAsErrors": true,
+    "keyFile": "../../tools/Key.snk"
+  },
+  "shared": "**/*.cs",
```
Comment (davidfowl) [inline src/Microsoft.Extensions.CommandLineUtils.Sources/project.json]: I think we should un-internalize this.

**Comment by BrennanConroy** [comment]
So non-sources package and un-internalize?

**Comment by muratg** [comment]
Yeah, regular package, un-internalize.

**Comment by davidfowl** [comment]
YAr

**Comment by BrennanConroy** [comment]
:up: :date: 
Moved to Microsoft.Extension.CommandLineUtils namespace
Made some classes public
Removed shared from project.json

**Comment by pranavkm** [comment]
Add the package to `NuGetPackageVerifier.json`.

:shipit:

**Pull request merged**